frontend/lib/resolution/prims.cpp

```diff
@@ -511,6 +526,255 @@ static QualifiedType primObjectToInt(Context* context, const CallInfo& ci) {
+static QualifiedType primIsGenericType(Context* context, const CallInfo& ci) {
+  if (ci.numActuals() < 1) return {};
+  bool eval = false;
+  auto qt = ci.actual(0).type();
+  if (auto t = qt.type()) {
+    auto g = t->genericity();
+
+    if (g == Type::MAYBE_GENERIC) {
+      QualifiedType qtEval = qt;
+      auto ct = t->toCompositeType();
+      if (auto cls = t->toClassType()) ct = cls->basicClassType();
+
+      // A generic type with defaults will be computed to have default subs
+      // at this point, so calling 'getTypeGenericity' on it will produce
+      // 'CONCRETE' instead of 'GENERIC_WITH_DEFAULTS'. Back up and use the
+      // uninstantiated type to produce the latter.
+      if (ct) {
+        if (auto ctGeneric = ct->instantiatedFromCompositeType()) {
+          qtEval = QualifiedType(qt.kind(), ctGeneric);
+        }
+      }
+      g = getTypeGenericity(context, qtEval);
+    }
+
+    // Both cases are considered 'generic' for this primitive.
+    eval = (g == Type::GENERIC || g == Type::GENERIC_WITH_DEFAULTS);
+  }
+  return makeParamBool(context, eval);
+}
+
+// TODO: No to 'FLAG_C_PTR_CLASS', 'FLAG_DATA_CLASS', 'FLAG_REF'.
+static QualifiedType primIsClassType(Context* context, const CallInfo& ci) {
+  if (ci.numActuals() < 1) return {};
+  bool eval = false;
+  if (auto t = ci.actual(0).type().type()) {
+    // TODO: Is it possible for the second and third to appear at all?
+    eval = t->isClassType()      ||  // E.g., 'owned C'.
+           t->isBasicClassType() ||  // E.g., 'C' in 'owned C'.
+           t->isAnyClassType();      // E.g., 'owned class'.
+  }
+  return makeParamBool(context, eval);
+}
+
+static QualifiedType
+primIsNilableClassType(Context* context, const CallInfo& ci) {
+  if (ci.numActuals() < 1) return {};
+  bool eval = false;
+  if (auto t = ci.actual(0).type().type()) {
+    if (auto mt = t->toClassType()) eval = mt->decorator().isNilable();
+  }
+  return makeParamBool(context, eval);
+}
+
+static QualifiedType
+primIsNonNilableClassType(Context* context, const CallInfo& ci) {
+  if (ci.numActuals() < 1) return {};
+  bool eval = false;
+  if (auto t = ci.actual(0).type().type()) {
+    if (auto mt = t->toClassType()) eval = mt->decorator().isNonNilable();
+  }
+  return makeParamBool(context, eval);
+}
+
+static QualifiedType primIsRecordType(Context* context, const CallInfo& ci) {
+  if (ci.numActuals() < 1) return {};
+  bool eval = false;
+  if (auto t = ci.actual(0).type().type()) eval = t->isUserRecordType();
+  return makeParamBool(context, eval);
+}
+
+static QualifiedType primIsFcfType(Context* context, const CallInfo& ci) {
+  CHPL_UNIMPL("PRIM_IS_FCF_TYPE");
+  return {};
+}
+
+static QualifiedType primIsUnionType(Context* context, const CallInfo& ci) {
+  if (ci.numActuals() < 1) return {};
+  bool eval = false;
+  if (auto t = ci.actual(0).type().type()) eval = t->isUnionType();
+  return makeParamBool(context, eval);
+}
```

Comment:
A lot of code duplication here. You could do away with a big chunk of it using a `typePredicate` helper:

```C++
template <typename Pred>
static QualifiedType typePredicate(Context* context, const CallInfo& ci, Pred&& pred) {
  if (ci.numActuals() != 1) return QualifiedType();
  bool eval = false;
  if (auto t = ci.actual(0).type().type()) eval = pred(t);
  return makeParamBool(context, eval);
}

static QualifiedType
primIsNilableClassType(Context* context, const CallInfo& ci) {
  return typePredicate(context, ci, [](const Type* t) {
    if (auto mt = t->toClassType()) return mt->decorator().isNilable();
    return false;
  });
}

static QualifiedType
primIsNonNilableClassType(Context* context, const CallInfo& ci) {
  return typePredicate(context, ci, [](const Type* t) {
    if (auto mt = t->toClassType()) return mt->decorator().isNonNilable();
    return false;
  });
}

static QualifiedType primIsRecordType(Context* context, const CallInfo& ci) {
  return typePredicate(context, ci, [](const Type* t) {
    return t->isUserRecordType();
  });
}

static QualifiedType primIsFcfType(Context* context, const CallInfo& ci) {
  CHPL_UNIMPL("PRIM_IS_FCF_TYPE");
  return {};
}

static QualifiedType primIsUnionType(Context* context, const CallInfo& ci) {
  return typePredicate(context, ci, [](const Type* t) {
    return t->isUnionType();
  });
}
```